fv3_cap.F90

```diff
@@ -0,0 +1,1691 @@
+!
+           call ESMF_FieldBundleRegrid(fcstFB(i), wrtFB(i,n_group),         &
+                                       routehandle=routehandle(i, n_group), &
+                                       termorderflag=(/ESMF_TERMORDER_SRCSEQ/), rc=rc)
+           if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+!
+!end FBcount
+         enddo
+         call ESMF_VMEpochExit(rc=rc)
+         if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+         timerh = mpi_wtime()
+         if (mype == 0 .or. mype == lead_wrttask(n_group)) print *,'aft fieldbundleregrid,na=',na,  &
+           ' time=', timerh- timerhi
+
+!      if(mype==0 .or. mype==lead_wrttask(1))  print *,'on wrt bf wrt run, na=',na
+          call ESMF_LogWrite('Model Advance: before wrtcomp run ', ESMF_LOGMSG_INFO, rc=rc)
+          if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+          timerhi = mpi_wtime()
+          call ESMF_GridCompRun(wrtComp(n_group), importState=wrtState(n_group), clock=clock_fv3,userRc=urc,rc=rc)
+          timerh = mpi_wtime()
+          if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+          if (ESMF_LogFoundError(rcToCheck=urc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__, rcToReturn=rc)) return
+!       if (mype == 0 .or. mype == lead_wrttask(n_group)) print *,'aft wrtgridcomp run,na=',na,  &
+!        ' time=', timerh- timerhi
+
+          call ESMF_LogWrite('Model Advance: after wrtcomp run ', ESMF_LOGMSG_INFO, rc=rc)
+          if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+!       if (mype == 0 .or. mype == lead_wrttask(n_group)) print *,'fv3_cap,aft model advance,na=', &
+!       na,' time=', mpi_wtime()- timewri
+
+
+          if(n_group == write_groups) then
+            n_group = 1
+          else
+            n_group = n_group + 1
+          endif
+
+        endif output
+
+! end quilting
+      endif
+
+!      if (mype == 0 .or. mype == 1536 .or. mype==2160) then
+!        print *,'fv3_cap,end integrate,na=',na,' time=',mpi_wtime()- timewri
+!      endif
+
+!*** end integreate loop
+    enddo integrate
+!
+!jw for coupled, check clock and dump import and export state
+    if ( cpl ) then
+      if( dbug > 0 .or. cplprint_flag ) then
+          call diagnose_cplFields(gcomp, importState, exportstate, clock_fv3,    &
+                                  fcstpe, cplprint_flag, dbug, 'export', export_timestr)
+     end if
+    endif
+
+    if (mype==0) print *,'fv3_cap,end integrate,na=',na,' time=',mpi_wtime()- timeri
+
+    if (profile_memory) call ESMF_VMLogMemInfo("Leaving FV3 Model_ADVANCE: ")
+
+  end subroutine ModelAdvance
+
+  !-----------------------------------------------------------------------------
+
+  subroutine ModelAdvance_phase1(gcomp, rc)
```

Comment:
If ModelAdvance is a full advance, then it should call ModelAdvance_phase1 and ModelAdvance_phase2 to reuse that code.